### ISSUE TITLE
comma separated bytesize format?

### ISSUE THREAD

**Issue by Ionian316** [MehdiK/Humanizer#426]
Is there a easy way to convert a number to a comma separated KB string? So that

32768 => "32 KB" and
0 => "0 KB" and
10000000 => "9,765 KB" ?

The only way I can achieve this right now is using
```c#
int i = 32768;
string s = i.Bytes().Kilobytes.ToString("N0") + " KB");
```
Is there a simpler way?

**Comment by MehdiK** [comment]
Haven't tried it but `(32768).Kilobytes().Humanize("{0:n0}")` might give you what you want?

**Comment by Ionian316** [comment]
Using your code gives me **{3:n2} MB** for 32768 and **{0:n0} b** for 0. I want to show **32 KB** and **0 KB** respectively. I also need to show a comma if the number needs a place separator.

**Comment by MehdiK** [comment]
You're absolutely right. My recollection was that the format gets passed directly to `String.Format` but I was wrong. This could be quite useful. I'll leave the issue open so we can tackle it.

We do accept PRs too :)

**Comment by jeremysimmons** [comment]
@Ionian316 
There is a difference between Int32.Format(string) and System.String(string, object[])
```csharp
int i = 10;
i.ToString("{0}") => {10}
String.Format("{0}", i" => 10
```

The Humanize extension methods generally follow the Int32.ToString style, rather than the more explicit String.Format style

The code you may want is
```csharp
int i = 32768;
string s = i.Bytes().Humanize("0 KB"); => 32 KB
int zero = 0;
zero.Bytes().Humanize("# KB"); => ' KB' // no unit displayed. This IS documented (see link)
zero.Bytes().Humanize("0 KB"); => 0 KB // what you really want!
zero.Bytes().Humanize("KB"); => 0 KB // I am pushing a fix for this
zero.Bytes().Humanize(); => 0 b // if you don't define a format, you get the bits format string...
```
refer to [Custom Numeric Format Strings](https://msdn.microsoft.com/en-us/library/0c899ak8(v=vs.110).aspx) documentation from microsoft.

**Comment by jeremysimmons** [comment]
@MehdiK  what do you think about chaging the default format to something like 
###,###,###,##0.## to handle more common scenarios?

The custom #, 0, and , specifiers are documented here:
https://msdn.microsoft.com/en-us/library/0c899ak8(v=vs.110).aspx
here is the tl;dr; version:
0 always prints
# prints if necessary
, is the localized group placeholder and only prints if a digit from # or 0 is printed on both sides
. is the localized decimal placedholder

**Comment by MehdiK** [comment]
I don't think this is solved by #440 @jeremysimmons. So I'll reopen it. I like your suggestion above though and that will fix the issue.

**Comment by jeremysimmons** [comment]
will send a PR implementing the suggestion later today

**Comment by jeremysimmons** [comment]
I remember now why I didn't fix the case of adding the separator 
There was a specific bug with the original format #.## that would prevent 0 from ever displaying. 
Changing the default format to include the placeholder separator is going to be a breaking change in the output. Someone may be depending on the current length of the output for formatting or later parsing. The current output matches more closely the default output format of G 1234.5678 ("G", en-US) -> 1234.56, the default standard numeric format. The requested change will be more like N 1234.5678 ("N", en-US) -> 1,234.57 which includes the placeholders.

The question originally asked by @Ionian316 can be resolved by using a specific format for that case.
Is it worth breaking compatibility?

**Comment by MehdiK** [comment]
Thanks @jeremysimmons. No breaking changes in minor versions please.

We could potentially add that as a configuration point (e.g. `Configurator.ByteSizeDefaultFormat`) and default it to the current value. 

I'm wondering if we should instead just pass the provided format through to `String.Format` (although I think that might be rather complex). That way the user is in full (granular) control of how they want each call to be formatted and not be locked into one format for ALL their ByteSize calls. Thoughts?

**Comment by jeremysimmons** [comment]
Acknowledged on the policy of no breaking changes. 
Output formatting is a breaking change. 

Adding a field to `Configurator` is a good idea, as is adding a second optional argument to Humanize/ToString.

My first reply to Ionian316 described the difference between Int32.Format and String.Format

You cannot pass the format to both String.Format and Int32.Format (double.format in the case of ByteSize) and get the same result.

They're different by nature.

```
int i = 10;
i.ToString("{0}") => {10}
String.Format("{0}", i" => 10
```

I think this is an education and documentation issue. Not a feature issue.

**Comment by Ionian316** [comment]
@jeremysimmons Thanks for your detailed explanation. I didn't realize the difference between the two formatters. `i.Bytes().Humanize("###,###,##0 KB")` did the trick for me.

**Comment by MehdiK** [comment]
The fix for this is now released to NuGet as v1.37.7.
